target/linux/realtek/files-6.6/drivers/i2c/busses/i2c-gpio-shared.c

```diff
@@ -0,0 +1,177 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Bitbanging driver for multiple I2C busses with shared SCL pin using the GPIO API
+ * Copyright (c) 2025 Markus Stockhausen <markus.stockhausen at gmx.de>
+ */
+
+#include <linux/i2c-algo-bit.h>
+#include <linux/gpio/consumer.h>
+#include <linux/mod_devicetable.h>
+#include <linux/mutex.h>
+#include <linux/platform_device.h>
+
+#define SI2C_MAX_BUS	4
+
+struct si2c_ctx;
+
+struct si2c_bus {
+	int nr;
+	struct gpio_desc *sda;
+	struct i2c_adapter adap;
+	struct i2c_algo_bit_data bit_data;
+	struct si2c_ctx *ctx;
+};
+
+struct si2c_ctx {
+	struct device *dev;
+	struct gpio_desc *scl;
+	struct mutex lock;
+	struct si2c_bus bus[SI2C_MAX_BUS];
```

Comment:
Why not dynamically allocate the bus children?

Reply:
Because our finally upstreamed RTL9300 I2C driver does the same. https://elixir.bootlin.com/linux/v6.15-rc6/source/drivers/i2c/busses/i2c-rtl9300.c